### PR TITLE
first encounters commit

### PR DESCRIPTION
**Pull request by ThyWoof**
. migrated custom encounters and heroes AI from DMPRO

### CODE REVIEW

**Comment by ThyWoof** [comment]
@PhilPJL , I haven't fully checked the encounter viewer code for any code smells

File: SolastaCommunityExpansion/Models/EncounterSpawnContext.cs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+            return Heroes;
+        }
+
+        internal static void ConfirmStageEncounter()
+        {
+            var position = GetEncounterPosition();
+
+            Gui.GuiService.ShowMessage(
+                MessageModal.Severity.Attention2,
+                "Encounter", $"Spawn encounter at coordinates ({position.x}, {position.z})?",
```
Comment (PhilPJL) [inline SolastaCommunityExpansion/Models/EncounterSpawnContext.cs:110]: do these need localization?

File: SolastaCommunityExpansion/Patches/HeroController/GameLocationBattlePatcher.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using HarmonyLib;
+
+namespace SolastaCommunityExpansion.Patches.HeroController
+{
+    internal static class GameLocationBattlePatcher
+    {
+        [HarmonyPatch(typeof(GameLocationBattle), "CheckVictoryOrDefeat")]
+        [SuppressMessage("Minor Code Smell", "S101:Types should be named in PascalCase", Justification = "Patch")]
+        internal static class GameLocationBattleCheckVictoryOrDefeat
```
Comment (PhilPJL) [inline SolastaCommunityExpansion/Patches/HeroController/GameLocationBattlePatcher.cs:10]: GameLocationBattle_CheckVictoryOrDefeat

**Review by PhilPJL — CHANGES_REQUESTED** [review]
See comments

**Review by PhilPJL — APPROVED** [review]

**Pull request merged**